File: scripts/merge_batch_patches.py
"""Merge per-batch judgment patches into a single judgment_patches.json."""

from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

if hasattr(sys.stdin, 'reconfigure'):
    sys.stdin.reconfigure(encoding='utf-8')
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')
if hasattr(sys.stderr, 'reconfigure'):
    sys.stderr.reconfigure(encoding='utf-8')


def merge_patches(batch_dir: str, scored_semi_path: str) -> dict:
    """Merge all patches_*.json files from batch_dir into one judgment_patches structure."""
    batch_path = Path(batch_dir)
    patch_files = sorted(batch_path.glob("patches_*.json"))

    if not patch_files:
        print("FATAL: No patches_*.json files found in batch directory", file=sys.stderr)
        sys.exit(1)

    merged_patches = {}
    schema_version = None
    model_version = "unknown"

    for pf in patch_files:
        with open(pf, encoding="utf-8") as f:
            raw = f.read()
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            start = raw.find("{")
            end = raw.rfind("}")
            if start >= 0 and end > start:
                data = json.loads(raw[start:end + 1])
            else:
                print(f"FATAL: Cannot parse JSON from {pf.name}", file=sys.stderr)
                sys.exit(1)

        sv = data.get("schema_version")
        if schema_version is None:
            schema_version = sv
        elif sv != schema_version:
            print(f"WARNING: schema_version mismatch in {pf.name}: "
                  f"'{sv}' vs '{schema_version}'", file=sys.stderr)

        mv = data.get("model_version", "unknown")
        if mv != "unknown":
            model_version = mv

        patches = data.get("patches", {})
        for rule_id, patch_data in patches.items():
            if rule_id in merged_patches:
                print(f"WARNING: Duplicate rule {rule_id} across batches, "
                      f"last one wins (from {pf.name})", file=sys.stderr)
            merged_patches[rule_id] = patch_data

    with open(scored_semi_path, encoding="utf-8") as f:
        scored_data = json.load(f)

    expected_ids = {r["id"] for r in scored_data.get("rules", []) if "id" in r}
    merged_ids = set(merged_patches.keys())

    missing = expected_ids - merged_ids
    if missing:
        print(f"WARNING: {len(missing)} rule IDs missing after merge: "
              f"{', '.join(sorted(missing)[:10])}", file=sys.stderr)

    extra = merged_ids - expected_ids
    if extra:
        print(f"WARNING: {len(extra)} unexpected rule IDs in patches: "
              f"{', '.join(sorted(extra)[:10])}", file=sys.stderr)

    return {
        "schema_version": schema_version or "0.1",
        "model_version": model_version,
        "patches": merged_patches,
    }


def main():
    output_path = None
    positional = []
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        if args[i] == "--output" and i + 1 < len(args):
            output_path = args[i + 1]
            i += 2
        else:
            positional.append(args[i])
            i += 1

    if len(positional) < 2:
        print("Usage: merge_batch_patches.py <batch_dir> <scored_semi.json> [--output file]",
              file=sys.stderr)
        sys.exit(1)

    batch_dir = positional[0]
    scored_semi_path = positional[1]

    result = merge_patches(batch_dir, scored_semi_path)
    if output_path:
        with open(output_path, "w", encoding="utf-8") as f:
            json.dump(result, f, indent=2, ensure_ascii=False)
    else:
        json.dump(result, sys.stdout, indent=2, ensure_ascii=False)
        sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: scripts/merge_batch_patches.js
"use strict";

const fs = require("fs");
const path = require("path");

function mergePatches(batchDir, scoredSemiPath) {
  const patchFiles = fs
    .readdirSync(batchDir)
    .filter((f) => /^patches_.*\.json$/.test(f))
    .sort()
    .map((f) => path.join(batchDir, f));

  if (patchFiles.length === 0) {
    console.error("FATAL: No patches_*.json files found in batch directory");
    process.exit(1);
  }

  const mergedPatches = {};
  let schemaVersion = null;
  let modelVersion = "unknown";

  for (const pf of patchFiles) {
    const raw = fs.readFileSync(pf, "utf-8");
    let data;
    try {
      data = JSON.parse(raw);
    } catch {
      const start = raw.indexOf("{");
      const end = raw.lastIndexOf("}");
      if (start >= 0 && end > start) {
        try {
          data = JSON.parse(raw.slice(start, end + 1));
        } catch {
          console.error(`FATAL: Cannot parse JSON from ${path.basename(pf)}`);
          process.exit(1);
        }
      } else {
        console.error(`FATAL: Cannot parse JSON from ${path.basename(pf)}`);
        process.exit(1);
      }
    }

    const sv = data.schema_version;
    if (schemaVersion === null) {
      schemaVersion = sv;
    } else if (sv !== schemaVersion) {
      console.error(
        `WARNING: schema_version mismatch in ${path.basename(pf)}: '${sv}' vs '${schemaVersion}'`
      );
    }

    const mv = data.model_version ?? "unknown";
    if (mv !== "unknown") {
      modelVersion = mv;
    }

    const patches = data.patches ?? {};
    for (const [ruleId, patchData] of Object.entries(patches)) {
      if (ruleId in mergedPatches) {
        console.error(
          `WARNING: Duplicate rule ${ruleId} across batches, last one wins (from ${path.basename(pf)})`
        );
      }
      mergedPatches[ruleId] = patchData;
    }
  }

  const scoredData = JSON.parse(fs.readFileSync(scoredSemiPath, "utf-8"));

  const expectedIds = new Set(
    (scoredData.rules ?? []).filter((r) => "id" in r).map((r) => r.id)
  );
  const mergedIds = new Set(Object.keys(mergedPatches));

  const missing = [...expectedIds].filter((id) => !mergedIds.has(id));
  if (missing.length) {
    console.error(
      `WARNING: ${missing.length} rule IDs missing after merge: ${missing.sort().slice(0, 10).join(", ")}`
    );
  }

  const extra = [...mergedIds].filter((id) => !expectedIds.has(id));
  if (extra.length) {
    console.error(
      `WARNING: ${extra.length} unexpected rule IDs in patches: ${extra.sort().slice(0, 10).join(", ")}`
    );
  }

  return {
    schema_version: schemaVersion || "0.1",
    model_version: modelVersion,
    patches: mergedPatches,
  };
}

function main() {
  let outputPath = null;
  const positional = [];
  const args = process.argv.slice(2);
  let i = 0;
  while (i < args.length) {
    if (args[i] === "--output" && i + 1 < args.length) {
      outputPath = args[i + 1];
      i += 2;
    } else {
      positional.push(args[i]);
      i += 1;
    }
  }

  if (positional.length < 2) {
    console.error("Usage: merge_batch_patches.js <batch_dir> <scored_semi.json> [--output file]");
    process.exit(1);
  }

  const batchDir = positional[0];
  const scoredSemiPath = positional[1];

  const result = mergePatches(batchDir, scoredSemiPath);
  const text = JSON.stringify(result, null, 2);
  if (outputPath) {
    fs.mkdirSync(path.dirname(outputPath), { recursive: true });
    fs.writeFileSync(outputPath, text, "utf-8");
  } else {
    process.stdout.write(text);
    process.stdout.write("\n");
  }
}

if (require.main === module) {
  main();
}

module.exports = { mergePatches };
